File: Version0.1/Python/Preprocessing/mkmaster.py
from numpy import *
from masternodes import *
from gaussnodes import *
# import importlib
# import mkshape #import the module here, so that it can be reloaded.
# importlib.reload(mkshape)
from mkshape import *

def mkmaster(dim,porder,pgauss,elemtype,nodetype):

    xpe,telem,xpf,tface,perm = masternodes(porder,dim,elemtype)

    gpe, gwe = gaussnodes(pgauss,dim,elemtype)
    gpe = array(gpe,float)
    gwe = array(gwe,float)
    gwe = reshape(gwe,(gwe.size,1))
    # shape functions and derivatives on the master volume element
    shapeg = mkshape(porder,xpe,gpe,elemtype)
    shapen = mkshape(porder,xpe,xpe,elemtype)

    npe     = shapeg.shape[0];
    nge     = shapeg.shape[1];
    nd1     = shapeg.shape[2];
    shapegt = zeros((nge,npe,nd1))
    shapent = zeros((npe,npe,nd1))
    shapegw = zeros((npe,nge,nd1))
    Mgwe = zeros((nge,nge))
    for i in range(0,nge):
        Mgwe[i,i] = gwe[i]

    for d in range(0,nd1):
        shapegt[:,:,d] = shapeg[:,:,d].transpose()
        shapent[:,:,d] = shapen[:,:,d].transpose()
        shapegw[:,:,d] = dot(shapeg[:,:,d],Mgwe)

    if dim>1:
        # Gauss points and weights on the master face element
        gpf, gwf = gaussnodes(pgauss,dim-1,elemtype)
        # shape functions and derivatives on the master face element
        shapfg = mkshape(porder,xpf,gpf,elemtype)
        shapfn = mkshape(porder,xpf,xpf,elemtype)
    else:
        gpf   = [0.0]
        gwf   = [1.0]
        shapfg =[1.0]
        shapfn =[1.0]

    npf = shapfg.shape[0];
    ngf = shapfg.shape[1];
    shapfgt = zeros((ngf,npf,dim));
    shapfnt = zeros((npf,npf,dim));
    shapfgw = zeros((npf,ngf,dim));
    Mgwf = zeros((ngf,ngf));
    for i in range(0,ngf):
        Mgwf[i,i] = gwf[i];
    for d in range(0,dim):
        shapfgt[:,:,d] = shapfg[:,:,d].transpose();
        shapfnt[:,:,d] = shapfn[:,:,d].transpose();
        shapfgw[:,:,d] = dot(shapfg[:,:,d],Mgwf);

    xp1d = masternodes(porder,1,elemtype)[0]
    gp1d,gw1d = gaussnodes(pgauss,1,elemtype)
    shap1dg = mkshape(porder,xp1d,gp1d,elemtype)
    shap1dn = mkshape(porder,xp1d,xp1d,elemtype)

    np1d = shap1dg.shape[0]
    ng1d = shap1dg.shape[1]
    shap1dgt = zeros((ng1d,np1d,2))
    shap1dnt = zeros((np1d,np1d,2))
    shap1dgw = zeros((np1d,ng1d,2))
    Mgw1d = zeros((ng1d,ng1d))
    for i in range(0,ng1d):
        Mgw1d[i,i] = gw1d[i]
    for d in range(0,2):
        shap1dgt[:,:,d] = shap1dg[:,:,d].transpose()
        shap1dgw[:,:,d] = dot(shap1dg[:,:,d],Mgw1d)
        shap1dnt[:,:,d] = shap1dn[:,:,d].transpose()

    master = {'nd' : dim, 'porder' : porder, 'pgauss' : pgauss, 'elemtype' : elemtype,
              'nodetype' : nodetype, 'npe' : npe, 'npf' : npf, 'nge' : nge, 'ngf' : ngf,
              'shapegt' : shapegt, 'shapegw' : shapegw, 'shapfgt' : shapfgt, 'shapfgw' : shapfgw,
              'shapent' : shapent, 'shapen' : shapen, 'shapfnt' : shapfnt, 'shapfn' : shapfn,
              'xpe' : xpe, 'gpe' : gpe, 'gwe' : gwe, 'xpf' : xpf, 'gpf' : gpf, 'gwf' : gwf,
              'shap1dgt' : shap1dgt, 'shap1dgw' : shap1dgw, 'shap1dnt' : shap1dnt,
              'shap1dn' : shap1dn, 'xp1d' : xp1d, 'gp1d' : gp1d, 'gw1d' : gw1d,
              'telem' : telem, 'tface' : tface, 'perm' : perm};

    return master

    # master = {'dim' : [], 'porder' : [], 'pgauss' : [], 'elemtype' : [], 'nodetype' : [],
    #           'npe' : [], 'npf' : [], 'nge' : [], 'ngf' : [], 'shapegt' : [], 'shapegw' : [],
    #           'shapfgt' : [], 'shapfgw' : [], 'shapent' : [], 'shapen' : [], 'shapfnt' : [],
    #           'shapfn' : [], 'xpe' : [], 'gpe' : [], 'gwe' : [], 'xpf' : [], 'gpf' : [],
    #           'gwf' : [], 'shap1dgt' : [], 'shap1dgw' : [], 'shap1dnt' : [], 'shap1dn' : [],
    #           'xp1d' : [], 'gp1d' : [], 'gw1d' : [], 'telem' : [], 'tface' : []};
    #
    # master['dim'] = dim
    # master['porder'] = porder
    # master['pgauss'] = pgauss
    # master['elemtype'] = elemtype
    # master['nodetype'] = nodetype
    # master['xpe'] = xpe
    # master['telem'] = telem
    # master['xpf'] = xpf
    # master['tface'] = tface
    # master['perm'] = perm
    # master['shapen'] = shapen;
    # master['xp1d'] = xp1d
    # shapvl = mkshape(porder,plocal,gpvl,elemtype)
    # master['shapvl'] = shapvl
    #
    # if ndim>1:
    #     gpfc,gwfc = gaussquad(pgauss,ndim-1,elemtype)
    #     master['gpfc'] = gpfc
    #     master['gwfc'] = gwfc
    #
    #     shapfc = mkshape(porder,plocfc,gpfc,elemtype)
    #     master['shapfc'] = shapfc
    #     gpfc = array(gpfc,float)
    #     gwfc = array(gwfc,float)
    #     master['gpfc'] = gpfc
    #     master['gwfc'] = reshape(gwfc,(gwfc.size,1))
    # else:
    #     master['gpfc'] = 0.5
    #     master['gwfc'] = 1.0
    #     master['shapfc'] = 1.0
    #
    # npv = master['shapvl'].shape[0]
    # ngv = master['shapvl'].shape[1]
    # master['shapvgdotshapvl']  = zeros((npv*npv,ngv,ndim+1))
    # master['shapvt'] = zeros((ngv,npv,ndim+1))
    # master['shapvg'] = zeros((npv,ngv,ndim+1))
    # for d in range(0,ndim+1):
    #     master['shapvt'][:,:,d] = master['shapvl'][:,:,d].transpose()
    #     master['shapvg'][:,:,d] = dot(master['shapvl'][:,:,d],diag(master['gwvl'][:,0]))
    #     for ii in range(0,npv):
    #         for jj in range(0,npv):
    #             master['shapvgdotshapvl'][ii*npv+jj,:,d] = master['shapvg'][jj,:,d]*master['shapvl'][ii,:,0]
    #
    #
    # npf = master['shapfc'].shape[0]
    # ngf = master['shapfc'].shape[1]
    # master['shapfgdotshapfc']  = zeros((npf*npf,ngf,ndim))
    # master['shapft'] = zeros((ngf,npf,ndim))
    # master['shapfg'] = zeros((npf,ngf,ndim))
    # for d in range(0,ndim):
    #     master['shapft'][:,:,d] = master['shapfc'][:,:,d].transpose()
    #     master['shapfg'][:,:,d] = dot(master['shapfc'][:,:,d],diag(master['gwfc'][:,0]))
    #     for ii in range(0,npf):
    #         for jj in range(0,npf):
    #             master['shapfgdotshapfc'][ii*npf+jj,:,d] = master['shapfg'][jj,:,d]*master['shapfc'][ii,:,0]
    #
    # master['shapmv'] = master['shapvt']
    # master['shapmf'] = master['shapft']
    # master['shapmh'] = master['shapmf']
    # master['permgeom'] = master['perm']


# function mkmaster(dim::IntP,porder::IntP,pgauss::IntP,elemtype::IntP,nodetype::IntP)
#
# # node positions on the master element and master face
# xpe,telem,xpf,tface,perm = masternodes(porder,dim,elemtype);
#
# # Gauss points and weights on the master volume element
# gpe, gwe = gaussnodes(pgauss,dim,elemtype);
#
# # shape functions and derivatives on the master volume element
# shapeg = mkshape(porder,xpe,gpe,elemtype);
# shapen = mkshape(porder,xpe,xpe,elemtype);
#
# npe,nge,nd1 = size(shapeg);
# shapegt = zeros(nge,npe,nd1);
# shapent = zeros(npe,npe,nd1);
# shapegw = zeros(npe,nge,nd1);
# Mgwe = zeros(nge,nge);
# for i = 1:nge
#     Mgwe[i,i] = gwe[i];
# end
# for d=1:nd1
#     shapegt[:,:,d] = shapeg[:,:,d]';
#     shapent[:,:,d] = shapen[:,:,d]';
#     shapegw[:,:,d] = shapeg[:,:,d]*Mgwe;
# end
#
# if dim>1
#     # Gauss points and weights on the master face element
#     gpf, gwf = gaussnodes(pgauss,dim-1,elemtype);
#     # shape functions and derivatives on the master face element
#     shapfg = mkshape(porder,xpf,gpf,elemtype);
#     shapfn = mkshape(porder,xpf,xpf,elemtype);
# else
#     gpf   = [0.0];
#     gwf   = [1.0];
#     shapfg = reshape([1.0],1,1,1);
#     shapfn = reshape([1.0],1,1,1);
# end
#
# npf = size(shapfg,1);
# ngf = size(shapfg,2);
# shapfgt = zeros(ngf,npf,dim);
# shapfnt = zeros(npf,npf,dim);
# shapfgw = zeros(npf,ngf,dim);
# Mgwf = zeros(ngf,ngf);
# for i = 1:ngf
#     Mgwf[i,i] = gwf[i];
# end
# for d=1:dim
#     shapfgt[:,:,d] = shapfg[:,:,d]';
#     shapfnt[:,:,d] = shapfn[:,:,d]';
#     shapfgw[:,:,d] = shapfg[:,:,d]*Mgwf;
# end
#
# xp1d,~,~,~,~ = masternodes(porder,1,elemtype);
# gp1d,gw1d = gaussnodes(pgauss,1,elemtype);
# shap1dg = mkshape(porder,xp1d,gp1d,elemtype);
# shap1dn = mkshape(porder,xp1d,xp1d,elemtype);
#
# np1d = size(shap1dg,1);
# ng1d = size(shap1dg,2);
# shap1dgt = zeros(ng1d,np1d,2);
# shap1dnt = zeros(np1d,np1d,2);
# shap1dgw = zeros(np1d,ng1d,2);
# Mgw1d = zeros(ng1d,ng1d);
# for i = 1:ng1d
#     Mgw1d[i,i] = gw1d[i];
# end
# for d=1:2
#     shap1dgt[:,:,d] = shap1dg[:,:,d]';
#     shap1dgw[:,:,d] = shap1dg[:,:,d]*Mgw1d;
#     shap1dnt[:,:,d] = shap1dn[:,:,d]';
# end
#
# master = Structs.MasterStruct(dim, porder, pgauss, elemtype, nodetype, npe, npf,
#     nge, ngf, np1d, ng1d, perm, shapegt, shapegw, shapfgt, shapfgw, shapent, shapen,
#     shapfnt, shapfn, xpe, gpe, gwe, xpf, gpf, gwf, shap1dgt, shap1dgw, shap1dnt,
#     shap1dn, xp1d, gp1d, gw1d);
#
# return master;
#
# end
#
# function mkmaster(app)
#     master = mkmaster(app.nd,app.porder,app.pgauss,app.elemtype,app.nodetype);
#     return master;
# end
#
# end

File: Applications/NavierStokes/OAT15A/sources/exasim4.0/matlab/master/mkmaster.m
function master=mkmaster(mesh,pgauss)
%MKMASTER  Create master element structure
%    MASTER=MKMASTER(MESH)
%
%      MESH:      Mesh data structure
%      PGAUSS:    Degree of the polynomila to be integrated exactly
%                 (default: PGAUSS = 2*MESH.PORDER)
%

if nargin < 2 
    pgauss = max(2*mesh.porder,1); 
end

master.nd     = mesh.nd;       % problem dimension
master.porder = mesh.porder;   % polynomial degree for solution
if isfield(mesh,'morder')
    master.morder = mesh.morder;   % polynomial degree for geometry
else
    master.morder = mesh.porder;   % polynomial degree for geometry
end
% master.plocvl = mesh.plocal;   % nodal points on the master volume element
% master.plocfc = mesh.plocfc;   % nodal points on the master face element
% master.perm   = mesh.perm;     % locations of the nodal points on the faces of the master elements

dim      = mesh.nd;
elemtype = mesh.elemtype;      % element type flag: simplex or tensor
nodetype = mesh.nodetype;
porder   = mesh.porder;
morder   = master.morder;

% node positions on the master element
[master.plocvl,master.tlocvl,master.plocfc,master.tlocfc,permnode,permedge,permface] = masternodes(porder,dim,elemtype,nodetype);
if dim==1
    master.perm = permnode;
elseif dim==2
    master.perm = permedge;
elseif dim==3
    master.perm = permface;
end

% geometry nodes on the master element and face
[master.plocmv,master.tlocmv,master.plocmf,master.tlocmf,permnode,permedge,permface] = masternodes(morder,dim,elemtype,nodetype);
if dim==1
    master.permgeom = permnode;
elseif dim==2
    master.permgeom = permedge;
elseif dim==3
    master.permgeom = permface;
end

% Gauss points and weights on the master volume element  
[master.gpvl,master.gwvl] = gaussquad(pgauss,dim,elemtype);
% master.gpnvl = master.plocvl;
% master.gwnvl = (1/size(master.plocvl,1)) * ones(size(master.plocvl,1),1);

% shape functions and derivatives on the master volume element  
master.shapvl = mkshape(porder,master.plocvl,master.gpvl,elemtype);
% master.shapnvl = mkshape(porder,master.plocvl,master.plocvl,elemtype);

% shape functions and derivatives on the master volume element  
master.shapmv = mkshape(morder,master.plocmv,master.gpvl,elemtype);
master.shapmv = permute(master.shapmv,[2 1 3]);

if dim>1
    % Gauss points and weights on the master face element  
    [master.gpfc,master.gwfc] = gaussquad(pgauss,dim-1,elemtype);
%     master.gpnfc = master.plocfc;
%     master.gwnfc = (1/size(master.plocfc,1)) * ones(size(master.plocfc,1),1);

    % shape functions and derivatives on the master face element  
    master.shapfc = mkshape(porder,master.plocfc,master.gpfc,elemtype);
%     master.shapnfc = mkshape(porder,master.plocfc,master.plocfc,elemtype);

    % shape functions and derivatives on the master face element  
    master.shapmf = mkshape(morder,master.plocmf,master.gpfc,elemtype);
    master.shapmf = permute(master.shapmf,[2 1 3]);  
else
    master.plocfc = 0;
    master.gpfc   = 0;
    master.gwfc   = 1;
    master.shapfc = 1;
%     master.gpnfc   = 0;
%     master.gwnfc   = 1;
%     master.shapnfc = 1;
end

% mass and convection matrices on the master element
master.mass = squeeze(master.shapvl(:,:,1))*diag(master.gwvl)*squeeze(master.shapvl(:,:,1))';
for ii=1:dim
    master.conv(:,:,ii) = squeeze(master.shapvl(:,:,1))*diag(master.gwvl)*squeeze(master.shapvl(:,:,ii+1))';
end   

master.ngv = size(master.gpvl,1);   % number of gasss points per element
master.ngf = size(master.gpfc,1);   % number of gasss points per face
master.npv = size(master.plocvl,1); % number of nodal points per element
master.npf = size(master.plocfc,1); % number of nodal points per face

%--------------- update MASTER structure ----------------%
npv = master.npv;
ngv = master.ngv;
dim = master.nd;
master.shapvgdotshapvl  = zeros(npv*npv,ngv,dim+1);      
for d=1:dim+1 
    master.shapvt(:,:,d) = master.shapvl(:,:,d)';
    master.shapvg(:,:,d) = master.shapvl(:,:,d)*diag(master.gwvl);    
    for ii=1:npv
        for jj = 1:npv
            master.shapvgdotshapvl((ii-1)*npv+jj,:,d) = master.shapvg(jj,:,d).*master.shapvl(ii,:,1);                    
        end
    end            
end

npf = master.npf;
ngf = master.ngf;
% face shape functions and their derivatives 
master.shapfgdotshapfc  = zeros(npf*npf,ngf,dim);      
for d=1:dim
    master.shapft(:,:,d) = master.shapfc(:,:,d)';
    master.shapfg(:,:,d) = master.shapfc(:,:,d)*diag(master.gwfc);
    for ii=1:npf
        for jj = 1:npf
            master.shapfgdotshapfc((ii-1)*npf+jj,:,d) = master.shapfg(jj,:,d).*master.shapfc(ii,:,1);                    
        end
    end            
end


